### ISSUE TITLE
ClassCastException when having a Controller returning Parametrized Type

### ISSUE THREAD

**Issue by aromanet42** [jrcodeza/spring-openapi#48]
Hello,

I have a controller returning a Parametrized Type.

Let's say I have a DTO `PaginatedResponse`

```
public class PaginatedResponse<T> {
    private List<T> data;
    private int totalCount;

    public PaginatedResponse(List<T> data, int totalCount) {
        this.data = data;
        this.totalCount = totalCount;
    }

    public List<T> getData() {
        return data;
    }

    public int getTotalCount() {
        return totalCount;
    }
}
```

My controller returns `PaginatedResponse<Foo>`

```
@RestController
@RequestMapping
public static class ParametrizedResponseController {
    @GetMapping(path = "/foo", produces = "application/json")
    @Operation
    public PaginatedResponse<Foo> parametrizedReponse() {
        return null;
    }
}
```

When generating OpenAPI3 documentation, I have an error:

```
java.lang.ClassCastException: sun.reflect.generics.reflectiveObjects.TypeVariableImpl cannot be cast to java.lang.Class

	at com.github.jrcodeza.schema.generator.ComponentSchemaTransformer.getFieldSchema(ComponentSchemaTransformer.java:216)
	at com.github.jrcodeza.schema.generator.ComponentSchemaTransformer.lambda$getClassProperties$10(ComponentSchemaTransformer.java:187)
	at org.springframework.util.ReflectionUtils.doWithLocalFields(ReflectionUtils.java:694)
	at com.github.jrcodeza.schema.generator.ComponentSchemaTransformer.getClassProperties(ComponentSchemaTransformer.java:186)
	at com.github.jrcodeza.schema.generator.ComponentSchemaTransformer.transformSimpleSchema(ComponentSchemaTransformer.java:60)
	at com.github.jrcodeza.schema.generator.OpenAPIGenerator.scanSchemasFromPackages(OpenAPIGenerator.java:224)
	at com.github.jrcodeza.schema.generator.OpenAPIGenerator.createComponentsWrapper(OpenAPIGenerator.java:207)
	at com.github.jrcodeza.schema.generator.OpenAPIGenerator.generate(OpenAPIGenerator.java:120)
```

I think the problem is that the model scan is done is a dedicated Step. When scanning `PaginatedResponse`, the generator does not know which will be the value of the parametrized type `T`.

I have a suggestion: maybe the generator should scan models from Operations?
In `createRequestBody` and `createApiResponses`, we know exactly the type, with the value of the parametrized type. 
We could change:
```
MediaType mediaType = schemaGeneratorHelper.createMediaType(methodReturnType, null, getGenericParams(method));
```
by
```
MediaType mediaType = schemaGeneratorHelper.createMediaType(schemaGenerator.readClass(method.getGenericReturnType()));
```

The `schemaGenerator.readClass` would be a method analysing the given type, the same way the ComponentSchemaTransformer do it now.

Pros:
- in this step, we know exactly the type returned by the Operation
- the generated swagger will contain only entities used by Operation (while scanning whole packages, we have a lot of Objects that may not be used by Operations)

Cons:
- is it a breaking change? I don't know if the Java API will change, but the generated swagger will change for sure.
- it is a big refacto

I don't know if I am thinking the right way. Maybe it is the wrong direction?

**Comment by jrcodeza** [comment]
Hi @aromanet42 

is PaginatedResponse your class or some external library class? Current this responses are solved in SchemaGeneratorHelper line 71- you can see there ResponseEntity. 

I think we could in this case traverse also the root class + generic class. The difficulty can be if someone specifies here external library class, it can have several technical attributes which might not be wanted to be exposed.

If PaginatedResponse is your class, can't you list it under packagesToBeScanned?

**Comment by aromanet42** [comment]
Hello @jrcodeza 

Listing this class under packagesToBeScanned does not help me, because this class is declared with a generic type `PaginatedResponse<T>`.

When the SchemaGeneratorHelper scans it, the `T` type cannot be resolved.

The `T` type can only be resolved if we know how this class is used.

If their is two APIs, one with `PaginatedResponse<Foo>` and the other with `PaginatedResponse<Bar>`, I would like to have 2 "component schema", one for each.

example:

```
@Operation
public PaginatedResponse<Foo> parametrizedResponseFoo() {
...
}

@Operation
public PaginatedResponse<Bar> parametrizedResponseBar() {
...
}
```

I don't see how this could be done with the current system, because it scans the classes without the usage context. When the SchemaGeneratorHelper scans class, it does not know how the class will be used by APIs. It does not know that this `PaginatedResponse` will be used with `Foo` and `Bar` types.

I expect to have a result like this:

```json
{
  "openapi": "3.0.1",
  "info": {
    "title": "Test API",
    "description": "Test description",
    "version": "1.0.0"
  },
  "paths": {
    "/foo": {
      "get": {
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/PaginatedResponseFoo"
                }
              }
            }
          }
        }
      }
    },
    "/bar": {
      "get": {
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/PaginatedResponseBar"
                }
              }
            }
          }
        }
      }
    }
  },
  "components": {
    "schemas": {
      "PaginatedResponseFoo": {
        "type": "object",
        "properties": {
          "data": {
            "type": "array",
            "items": {
              "$ref": "#/components/schemas/Foo"
            }
          },
          "totalCount": {
            "type": "number"
          }
        }
      },
      "PaginatedResponseBar": {
        "type": "object",
        "properties": {
          "data": {
            "type": "array",
            "items": {
              "$ref": "#/components/schemas/Bar"
            }
          },
          "totalCount": {
            "type": "number"
          }
        }
      },
      "Foo": {
        "type": "object"
      },
      "Bar": {
        "type": "object"
      }
    }
  }
}

```

**Comment by jrcodeza** [comment]
@aromanet42 there is also Response annotation in spring-openapi. Doesn't this suit your need? You can define there explicitly in this case that you want to use PaginatedResponseBar for example. Or if you have some method or mechanism which merges generic type with it's wrapping type, then you can use OperationInterceptor. 

If this is not an option for you, I have to also think about possible automated solution. Or perhaps we can have a call, maybe I'm only not understanding everything the correct way :)

**Comment by aromanet42** [comment]
@jrcodeza 

I opened a PR adding a test to represent my need :)

https://github.com/jrcodeza/spring-openapi/pull/51

I just added a test, there is no code, no fix in there.

If you launch the test, you will see that it fail with `java.lang.ClassCastException: sun.reflect.generics.reflectiveObjects.TypeVariableImpl cannot be cast to java.lang.Class`

I don't see how I could use the Response annotation in this case? You can of course edit the PR if you see a way to make it work ;)

**Comment by jrcodeza** [comment]
Hmm this is a more difficult scenario for generation. At the time of model lookup the type is unknown. It is defined at the time of usage in REST operation. You're right this must be implemented. Just thinking now how :)

**Comment by aromanet42** [comment]
In the mean time, I used the `io.swagger.v3.core.converter.ModelConverters` from swagger-core to handle this case.

It handle nicely this kind of use-case.

But I don't know how to integrate this in your library...

**Comment by jrcodeza** [comment]
I'm not sure if it should be integrated. The question is how then such component would be called in spec itself. Then if you generate client back from such spec then you either way loose the information about generics.

**Comment by aromanet42** [comment]
Hi @jrcodeza 
I understand your concerns. 
About the name, it could be the concatenation of the generic type and the actual type of the parameter. In my example, this would be `PaginatedResponseFoo`.

When generating client back from the spec, we will indeed loose the generics, but I don't think it is a problem. 

I managed to resolve my problem by reimplementing the way the Schemas are found. Sadly I had to stop using your library :(

Thanks a lot for your time and help
